arangod/Cluster/MaintenanceFeature.cpp

```diff
@@ -0,0 +1,442 @@
+  _secondsActionsBlock = 30;
+  _secondsActionsLinger = 300;
+
+  return;
+
+} // MaintenanceFeature::init
+
+
+void MaintenanceFeature::collectOptions(std::shared_ptr<ProgramOptions> options) {
+  options->addSection("server", "Server features");
+
+  options->addHiddenOption(
+    "--server.maintenance-threads",
+    "maximum number of threads available for maintenance actions",
+    new Int32Parameter(&_maintenanceThreadsMax));
+  
+  options->addHiddenOption(
+    "--server.maintenance-actions-block",
+    "minimum number of seconds finished Actions block duplicates",
+    new Int32Parameter(&_secondsActionsBlock));
+  
+  options->addHiddenOption(
+    "--server.maintenance-actions-linger",
+    "minimum number of seconds finished Actions remain in deque",
+    new Int32Parameter(&_secondsActionsLinger));
+
+} // MaintenanceFeature::collectOptions
+
+
+/// do not start threads in prepare
+void MaintenanceFeature::prepare() {
+} // MaintenanceFeature::prepare
+
+
+void MaintenanceFeature::start() {
+  int loop;
+  bool flag;
+
+  // start threads
+  for (loop=0; loop<_maintenanceThreadsMax; ++loop) {
+    maintenance::MaintenanceWorker * newWorker = new maintenance::MaintenanceWorker(*this);
+    _activeWorkers.push_back(newWorker);
+    flag = newWorker->start(&_workerCompletion);
+
+    if (!flag) {
+      LOG_TOPIC(ERR, Logger::CLUSTER)
+        << "MaintenanceFeature::start:  newWorker start failed";
+    } // if
+  } // for
+} // MaintenanceFeature::start
+
+
+void MaintenanceFeature::beginShutdown() {
+
+    _isShuttingDown=true;
+    CONDITION_LOCKER(cLock, _actionRegistryCond);
+    _actionRegistryCond.broadcast();
+
+    return;
+
+} // MaintenanceFeature
+
+
+void MaintenanceFeature::stop() {
+
+  for (auto itWorker : _activeWorkers ) {
+    CONDITION_LOCKER(cLock, _workerCompletion);
+
+    // loop on each worker, retesting at 10ms just in case
+    if (itWorker->isRunning()) {
+      _workerCompletion.wait(10000);
+    } // if
+  } // for
+
+  return;
+
+} // MaintenanceFeature::stop
+
+
+/// @brief Move an incomplete action to failed state
+Result MaintenanceFeature::deleteAction(uint64_t action_id) {
+  Result result;
+
+  // pointer to action, or nullptr
+  auto action = findActionId(action_id);
+
+  if (action) {
+    if (maintenance::COMPLETE != action->getState()) {
+      action->setState(maintenance::FAILED);
+    } else {
+      result.reset(TRI_ERROR_BAD_PARAMETER,"deleteAction called after action complete.");
+    } // else
+  } else {
+    result.reset(TRI_ERROR_BAD_PARAMETER,"deleteAction could not find action to delete.");
+  } // else
+
+  return result;
+
+} // MaintenanceFeature::deleteAction
+
+
+
+/// @brief This is the  API for creating an Action and executing it.
+///  Execution can be immediate by calling thread, or asynchronous via thread pool.
+///  not yet:  ActionDescription parameter will be MOVED to new object.
+Result MaintenanceFeature::addAction(
```

Comment:
None of the addAction() and createAction() routines explicitly check to see if construction of action set FAILED.  Therefore it is possible for an "executeNow" action to start running with known invalid parameters.